docs(blog): ibis, duckdb and lonboard for overture maps (#10143)
ibis-project/ibis: docs/posts/ibis-overturemaps/index.qmd

---
title: "From query to plot: Exploring GeoParquet Overture Maps with Ibis, DuckDB, and Lonboard"
author: Naty Clementi and Kyle Barron
date: 2024-09-13
categories:
  - blog
  - duckdb
  - overturemaps
  - lonboard
  - geospatial
---

With the release of `DuckDB 1.1.0`, now we have support for reading GeoParquet
files! With this exciting update we can query rich datasets from Overture Maps using python via Ibis with the performance of `DuckDB`.

But the good news don't stop there, since `Ibis 9.2`, `lonboard` can plot data directly from an `Ibis` table, adding more simplicity and speed to your geospatial analysis.

Let’s dive into how these tools come together.

## Installation

> Install Ibis with the dependencies needed to work with geospatial data using DuckDB. To be able to read geoparquet files the duckdb version should be `>=1.1.0`.

Reviewer: ```suggestion
Install Ibis with the dependencies needed to work with geospatial data using DuckDB. To be able to read GeoParquet files the DuckDB version should be `>=1.1.0`.
```

::: {.callout-note}

> At the moment duckdb 1.1.0 has a bug that prevents us from querying and writing the data to parquet using Ibis and DuckDB, so we are installing the `overturemaps` CLI to get the data.

Reviewer: ```suggestion
At the moment DuckDB 1.1.0 has a bug that prevents us from querying and writing the data to parquet using Ibis and DuckDB, so we are installing the `overturemaps` CLI to get the data.
```

:::

```bash
$ pip install 'ibis-framework[duckdb,geospatial]' lonboard overturemaps
```

## Motivation

> 

Reviewer: I'd start with a sentence intro about what Overture Maps is. Like

> Overture Maps is a project to build map data products on top of a variety of sources, like OpenStreetMap.

(not sure that's the best one-liner description of what Overture Maps is, any zingers @jwass ?)

> Overture Maps offers a variety of datasets to query, but we thought that it would
> be interesting to see some plots related to the power infrastructure. We'll look
> into power plants, and power lines of most of the USA (excluding territories and
> Alaska for simplicity of the bounding box).

Reviewer: ```suggestion
Overture Maps offers a variety of datasets to query. Let's create
some plots related to U.S. power infrastructure. We'll look
into power plants and power lines for the lower 48 states (excluding Hawaii and
Alaska for simplicity of the bounding box).
```

## Download data

**NOTE: not sure if I should have this here or avoid it because it's broken**

With Ibis and DuckDB we could download only the necessary columns needed but at
the moment this fails due to a bug.

For future reference when this gets fixed, the expression would look like this:

```python
import ibis
from ibis import _

con = ibis.get_backend()

# look into type infrastructure
url = "s3://overturemaps-us-west-2/release/2024-07-22.0/theme=base/type=infrastructure/*"
t = con.read_parquet(url, table_name="infra-usa")

# filter for USA bounding box, subtype="power", and selecting only few columns
expr = t.filter(_.bbox.xmin > -125.0, _.bbox.ymin > 24.8,

>                 _.bbox.xmax < -65.8, _.bbox.ymax < 49.2,

Reviewer: This is fine for this blog post, but it would also be nice if there was some way to put this into a `read_geoparquet` query so that the user doesn't have to reach into these `.bbox.xmin` columns themselves.

be interesting to see some plots related to the power infrastructure. We'll look
into power plants, and power lines of most of the USA (excluding territories and
Alaska for simplicity of the bounding box).

## Download data

**NOTE: not sure if I should have this here or avoid it because it's broken**

With Ibis and DuckDB we could download only the necessary columns needed but at
the moment this fails due to a bug.

For future reference when this gets fixed, the expression would look like this:

```python
import ibis
from ibis import _

con = ibis.get_backend()

# look into type infrastructure
url = "s3://overturemaps-us-west-2/release/2024-07-22.0/theme=base/type=infrastructure/*"
t = con.read_parquet(url, table_name="infra-usa")

# filter for USA bounding box, subtype="power", and selecting only few columns
expr = t.filter(_.bbox.xmin > -125.0, _.bbox.ymin > 24.8,
                _.bbox.xmax < -65.8, _.bbox.ymax < 49.2,
                _.subtype=="power"
                ).select(["names",
                          "geometry",
                          "class",
                          "sources",
                          "source_tags"])

```

::: {.callout-note}

> If you inspect expr, you can see that the filters and projections get pushed down
> meaning you only bring the data that you asked for.

Reviewer: ```suggestion
If you inspect expr, you can see that the filters and projections get pushed down,
meaning you only download the data that you asked for.
```

:::

```python
# to write it to a parquet file we would do (currently broken)
con.to_parquet(expr, "power-infra-usa.parquet")
```

> But, no worries! Overture Maps has a nice [python CLI](https://docs.overturemaps.org/getting-data/overturemaps-py/) that allow us to download

Reviewer: You can also do this from Python in case you wanted to avoid reaching for the command line, but this is fine.

Author: I just learned that the DuckDB bug fix will go out next week (see Max's comment below), so I'll rewrite this part to read download the data that way.

some the data, we can filter by bounding box and type but any other filters will have to happen afterwards.

```bash
$ overturemaps download --bbox=-125.0,24.8,-65.8,49.2 -f geoparquet --type=infrastructure -o usa-infra.geoparquet
```

Now that we have the data lets explore it in Ibis interactive mode and make some beautiful maps.

## Data exploration

```{python}
import ibis
from ibis import _

ibis.options.interactive = True

> con = ibis.get_backend() # default duckdb backend

Reviewer: I'm not sure the usual behavior in your demos, but especially since ibis just switched from Pandas to DuckDB as the default backend, I think it would be slightly clearer if this chose DuckDB explicitly.

Author: > Ibis just switched from Pandas to DuckDB  

This is not true, DuckDB has been the default backend for over 2 years now. I'll double check what do we do in other blogs to be more explicit.

Reviewer 2: DuckDB has been the default backend [since early 2022](https://github.com/ibis-project/ibis/commit/8ccb81d49e252b57310bdb3a97eeb77ef1d28bac)! 😂

Reviewer: Oh oops. I misunderstood your pandas deprecation as you just now switching to make that your default

Reviewer 3: It's the longest tenured default backend by a factor of +inf since it's the only default backend we've ever had.

Reviewer 2: Totally! I think this was a common perception from that blog, but I'm happy to have more places where there's a correction on the record.

Reviewer 2: I guess we all read and responded at effectively the same time 🤣

plants = usa_infra.filter(_.subtype=="power", usa_infra["class"]=="plant")
power_lines = usa_infra.filter(_.subtype=="power", usa_infra["class"]=="power_line")
minor_lines = usa_infra.filter(_.subtype=="power", usa_infra["class"]=="minor_line")
```

## Plotting

**Note: maybe here explain why lonboard ?**

::: {.callout-note}
You can try this in your machine, for the purpose the blog file size, we will show
screenshots of the visualization
:::

```python
import lonboard
from lonboard.basemap import CartoBasemap # to choose color of basemap
```

Let's visualize the `power plants`

```python
lonboard.viz(plants,
            scatterplot_kwargs={"get_fill_color": "red"},
            polygon_kwargs={"get_fill_color": "red"},
            map_kwargs={"basemap_style": CartoBasemap.Positron,
            "view_state": {"longitude": -100, "latitude": 36, "zoom": 3}
            })
```

![Power plants in the USA](usa-power-plants.png)

If you are visualizing this in your machine, you can zoom in and see some of the
geometry where the plants are located. But as an example, we can plot in a small
area of California:

```python
plants_CA = plants.filter(_.bbox.xmin.between(-118.6, -117.9),

>            _.bbox.ymin.between(34.5, 35.3))[_.names.primary, _.geometry]

Reviewer: Nit: I'd suggest using standard black/ruff code formatting for all Python code blocks

Author: Good catch, I'll fix this.

```

```python
lonboard.viz(plants_CA,
            scatterplot_kwargs={"get_fill_color": "red"},
            polygon_kwargs={"get_fill_color": "red"},
            map_kwargs={"basemap_style": CartoBasemap.Positron,
            })
```

> ![Power plants near Lancaster CA](ca-power-plants.png)

Reviewer: ```suggestion
![Power plants near Lancaster, CA](ca-power-plants.png)
```

We can also visualize together the `power_lines` and the `minor_lines` by doing:

```python
lonboard.viz([minor_lines, power_lines])
```

![Minor and Power lines of USA](usa-power-and-minor-lines.png)

and that's how you can visualize ~7 million points from the comfort of
your laptop.

> Note: I got the ~7M by adding the number of points in power_lines and minor_lines. I'm not sure if plotting the lines that connect these points add points to this.

Reviewer: It is more effort for the GPU to render lines than to render points, but a line isn't rendered as a collection of points.

It might be clearer to say `7 million coordinates` rather than points.

```python
>>> power_lines.geometry.n_points().sum()
5329836
>>> minor_lines.geometry.n_points().sum()
1430042
```

With Ibis and DuckDB working with geospatial data has never been easier or faster.
We saw how to query a dataset from Overture Maps with the simplicity of Python and

> the performance of DuckDB. Last but not least, we saw how simple and quick lonboard

Reviewer: nit: I try to regularly capitalize Lonboard because I've had people think it was `ionboard`
```suggestion
the performance of DuckDB. Last but not least, we saw how simple and quick Lonboard
```
